### PR TITLE
cli/cmd: implement kubeconfig fallback to Terraform state

### LINKED ISSUE #608: `lokoctl health` does not fallback for kubeconfig
I don't have assets dir locally but in S3 and when I run following: 

```console
$ lokoctl health
Error: Kubeconfig "assets/cluster-assets/auth/kubeconfig" not found
Usage:
  lokoctl health [flags]

Flags:
  -h, --help   help for health

Global Flags:
      --kubeconfig string     Path to kubeconfig file, taken from the asset dir if not given, and finally falls back to ~/.kube/config
      --lokocfg string        Path to lokocfg directory or file (default "./")
      --lokocfg-vars string   Path to lokocfg.vars file (default "./lokocfg.vars")
```

It did not fallback to `~/.kube/config` as mentioned in the `--kubeconfig` flag description.

And same for `component apply` as well:

```console
$ lokoctl component apply
Applying component 'openebs-operator'...
FATA[0000] stat assets/cluster-assets/auth/kubeconfig: no such file or directory  args="[]" command="lokoctl component apply"
```

### PR DESCRIPTION
**Pull request by invidian**
**This PR includes commits from #631, please review this one first and don't leave commit comments for this PR here.**

TODO:
- [x] Running `lokoctl health` is now slower, as on every run, it initializes the Terraform and reads the output. Should we try to read file from assets dir first and then fallback to Terraform output? Maybe we should cache somehow the `kubeconfig` file content?
- [x] Update and add new tests

Closes #608

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

### CODE REVIEW

**Comment by invidian** [comment]
> Running lokoctl health is now slower, as on every run, it initializes the Terraform and reads the output. Should we try to read file from assets dir first and then fallback to Terraform output? Maybe we should cache somehow the kubeconfig file content?

It seems caching is more complex than I thought, as the optimal sequence of execution is against our abstraction. What we need:
- Try loading platform configuration
- Check if kubeconfig file exists in platform assets directory (if it does, great, we're done
- initialize Terraform
- read Terraform output

And what we have right now:
- Try loading platform configuration
- initialize Terraform
- read Terraform output

**Comment by johananl** [comment]
>This PR includes commits from #631, please review this one first and don't leave commit comments for this PR here.

What do you mean by "this one" and "this PR"? #631 or #701? Should #701 be reviewed?

**Comment by invidian** [comment]
>What do you mean by "this one" and "this PR"? #631 or #701? Should #701 be reviewed?

It's not relevant anymore, as #631 has been merged already :) And yes, I requested reviews on this PR, so it should be reviewed.

File: cli/cmd/utils.go [hunk 1]
```diff
@@ -77,82 +83,88 @@ func getConfiguredPlatform() (platform.Platform, hcl.Diagnostics) {
...
+	// If no sources has been returned, it means we should read from Terraform state.
+	if len(sources) == 0 {
+		return readKubeconfigFromTerraformState(contextLogger)
 	}
 
-	if expandedPath, err := homedir.Expand(path); err == nil {
-		path = expandedPath
+	// Select first non-empty source and read it.
+	for _, source := range sources {
+		if source != "" {
```
Comment (knrt10) [inline cli/cmd/utils.go:102]: IIUC, sources will always have path string, so I think there is no need to check if it is empty or not

Comment (invidian) [reply cli/cmd/utils.go:102]: if `KUBECONFIG` environment variable is not set, sources will look like: `[]string{"", "~/.kube/config"}`, so this is needed IMO.

Comment (knrt10) [reply cli/cmd/utils.go:102]: Cool, get it

**Comment by invidian** [comment]
Rebased to resolve conflicts.

**Review by knrt10 — APPROVED** [review]
Looks great.

**Comment by surajssd** [comment]
This failed as we want:

```console
$ lokoctl component apply
WARN[0000] Kubeconfig file not found in assets directory, pulling kubeconfig from Terraform state, this might be slow. Run 'lokoctl cluster apply' to fix it.  args="[]" command="lokoctl component apply"
INFO[0004] Executing step "initialize Terraform"         phase=infrastructure

You can find the logs in "assets/terraform/logs/299669.log"
FATA[0005] Error in finding kubeconfig file: reading kubeconfig file content from Terraform state: failed getting Terraform output for key "kubeconfig": exit status 1  args="[]" command="lokoctl component apply"
```

```console
$ cat assets/terraform/logs/299669.log
Initializing modules...

Initializing the backend...

Initializing provider plugins...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```

Can we give more informational message to the user?

**Comment by invidian** [comment]
>Can we give more informational message to the user?

Hmm, should we replace top-level message (`Error in finding kubeconfig file: reading kubeconfig file content from Terraform state: failed getting Terraform output for key "kubeconfig": exit status 1`) entirely then?

Something like `Finding kubeconfig file failed. Did you run "lokoctl cluster apply" ?` ?

**Comment by surajssd** [comment]
@invidian that should do for now :-) It is better than the terraform one.

**Comment by invidian** [comment]
Maybe we can keep existing message in debug verbosity?

**Comment by surajssd** [comment]
Sounds ok @invidian

**Comment by invidian** [comment]
>Sounds ok @invidian

:confused: I had to add `--debug` flag, so debug messages can be actually displayed. I could just set it without the flag too, but I think it's a reasonable addition. And I would really prefer not to hide the error. Alternative would be to print exact error after friendly error message.

File: cli/cmd/component-apply.go [hunk 2]
```diff
@@ -41,6 +41,15 @@ When run with no arguments, all components listed in the configuration are appli
 	},
 }
 
+var debug bool
+
+// nolint:gochecknoinits
+func init() {
+	componentCmd.AddCommand(componentApplyCmd)
+	pf := componentApplyCmd.PersistentFlags()
```
Comment (knrt10) [inline cli/cmd/component-apply.go:49]: should we use 2 init functions? I am little confused, can you please give a little context?

Comment (invidian) [reply cli/cmd/component-apply.go:49]: Oh, we shouldn't, nice catch! (I realized now that you can have more than one `init()` function per package :exploding_head:

**Comment by knrt10** [comment]
Can you please update the docs again, it's failing because of the init function change

**Comment by invidian** [comment]
>Can you please update the docs again, it's failing because of the init function change

:facepalm: done :smile:

File: cli/cmd/health.go [hunk 3]
```diff
@@ -42,28 +45,33 @@ func runHealth(cmd *cobra.Command, args []string) {
...
-		contextLogger.Fatal("No cluster configured")
+	kubeconfig, err := getKubeconfig(contextLogger, lokoConfig, true)
+	if err != nil {
+		contextLogger.Debugf("Error in finding kubeconfig file: %s", err)
+		contextLogger.Fatal("Suitable kubeconfig file not found. Did you run 'lokoctl cluster apply' ?")
 	}
 
+	cs, err := k8sutil.NewClientset(kubeconfig)
+	if err != nil {
+		contextLogger.Fatalf("Error in creating setting up Kubernetes client: %q", err)
```
Comment (knrt10) [inline cli/cmd/health.go:69]: Seems little confusing.

```suggestion
		contextLogger.Fatalf("Error in setting up Kubernetes client: %q", err)
```

**Review by knrt10 — COMMENTED** [review]
Rest lgtm

File: cli/cmd/health.go (hunk 3 above)
Comment (invidian) [reply cli/cmd/health.go]: Out of scope of this PR, but I fixed it in separate commit.

**Review by knrt10 — APPROVED** [review]
+1

**Comment by surajssd** [comment]
This failed as follows:

```bash
$ lokoctl health --kubeconfig-file $HOME/.kube/config
ERRO[0000] <nil>: no platform configured;                args="[]" command="lokoctl health"
FATA[0000] Suitable kubeconfig file not found. Did you run 'lokoctl cluster apply' ?  args="[]" command="lokoctl health"

$ lokoctl version
v0.3.0-207-gf7f5eded9

$ ll
.rw-rw-r--@ 35 hummer  3 Sep 12:45 foobar.lokocfg

$ cat foobar.lokocfg 
component "prometheus-operator" {}
```

`lokoctl` still tries to get the file from state, even after explicit mention.

**Comment by invidian** [comment]
>lokoctl still tries to get the file from state, even after explicit mention.

This is as intended. `health` command should be only run on L8e clusters as far as I'm aware.

**Comment by surajssd** [comment]
Why do we have a `--kubeconfig-file` flag for this? Also the help says that it has a fallback mechanism, so it is confusing to the user. I understand the technicalities behind why it shows up here, but then for a user it can be confusing. Either we fail if we have a flag mentioned or this flag should not be here:

```
$ lokoctl health -h
Get the health of a cluster

Usage:
  lokoctl health [flags]

Flags:
      --debug   Print debug messages
  -h, --help    help for health

Global Flags:
      --kubeconfig-file string   Path to a kubeconfig file. If empty, the following precedence order is used:
                                   1. Cluster asset dir when a lokocfg file is present in the current directory.
                                   2. KUBECONFIG environment variable.
                                   3. ~/.kube/config file.
...
```

**Comment by invidian** [comment]
>Why do we have a --kubeconfig-file flag for this? Also the help says that it has a fallback mechanism, so it is confusing to the user. I understand the technicalities behind why it shows up here, but then for a user it can be confusing. Either we fail if we have a flag mentioned or this flag should not be here:

Because it's a global flag... :(

**Comment by invidian** [comment]
@surajssd I pushed the fix with moving `--kubeconfig-file` flag. Please have a look.

**Review by surajssd — APPROVED** [review]
Verified the `lokoctl health` functionality works.

**Review by knrt10 — APPROVED** [review]

**Pull request merged**